Add roles and permissions to the app

hanneskuettner opened the pull request:
<!--

Heya! Thanks for opening a Pull Request! If your PR is implementing a new feature or fix for Directus, please make sure your PR adheres to the following requirements:

- The PR closes an Issue (not Discussion)
- Tests are added/updated and are passing locally if applicable
- Documentation was added/updated if applicable

Please make sure to "Link" the issue you're closing. Without a Linked issue, this PR won't be accepted. See https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue for more information.

-->

## Scope


What's changed:

- The permissions table is now a system interface that can be used for attaching permissions to policies, this enables us to create policies in-line (in a drawer for that matter)
  - The permission details have been moved into that system interface as a separate drawer component
- Added policies to users and roles with the default M2M interface 

## Potential Risks / Drawbacks

- Only a not functioning app, no biggie

## Review Notes / Questions

## Todos

### Must haves

- [x] Permission store adaption to the new permissions system
  - [x] Use backend util endpoint for determining the merged permissions of a user
  - [x] Identifying usages of the old permissions
  - [x] Update usages of the old permissions
- [x] User with minimal app permissions can log in
- [x] Bring back public role configuration with policies instead of permissions
- [x] Decide which shares and presets a user with minimal and recommended app permissions can see, as it currently is a `_eq: $CURRENT_ROLE`
- [x] Decide if the `AppUser`, `ActionPermission` and `CollectionPermission` types should be moved to `@directus/types` or be duplicated in `@directus/extensions` for `defineModule`

### Could haves

- [ ] Display role tree in `roles/collection.vue`
- [ ] Show merged permissions in a custom m2m policies interface
- [ ] Show nested users under their parent role?

## (Incomplete) List of breaking changes

- The `app_access`, `admin_access` and `enforce_tfa` flags moved from `User.role` to the top level of a new `AppUser` that is used throughout the app
- The `preRegisterCheck` of a module is called with this new `AppUser` and a record of all user permissions in a condensed form (`Record<string, CollectionPermission>`) instead of a list of all permissions. To identify if a user has access to an action in a collection look for `permissions[collection]?.[action]`, which has the `access`, `full_access` and `fields` properties.

---

 Fixes #21764

hanneskuettner committed:
Fix getRelationsForField

hanneskuettner committed:
Add changeset

hanneskuettner committed:
Initial app changes

hanneskuettner committed:
Merge branch 'tmp' into auditus-app

hanneskuettner committed:
Remove app-permissions from role settings

hanneskuettner committed:
Make sure access row uuids are auto generated

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Move a few things around, set up policies m2m properly

hanneskuettner committed:
Merge remote-tracking branch 'origin/auditus-app' into auditus-app

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Show roles as tree in sidebar
Change avatar field query for user

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Show user and role count in policy table

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Default to not adding app access for a policy, makes composability less annoying

hanneskuettner committed:
Correctly fall back to 0 for counts

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Change the structure of current user permissions

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Start bringing back the public role

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Make the public role a real role rather than a virtual one

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Revert public role changes

DanielBiegler committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner commented on app/src/components/v-select/v-select.vue:
```diff
@@ -132,7 +132,7 @@ function useItems() {
 				selectable: get(item, props.itemSelectable),
 				children: children
 					? children.filter((childItem: Record<string, any>) =>
-							filterItem(get(childItem, props.itemText), get(childItem, props.itemValue), childItem.children),
```
This is an unrelated fix. `children` at this point is already parsed and transformed to a canonical item form, which now only has the `text` and `value` props, and not the custom ones that one could provide through `itemText` and `itemValue`.

hanneskuettner committed:
Extend list-m2m to allow for very custom junction matching and a primary key of `null`

hanneskuettner committed:
Remove unused

hanneskuettner committed:
Fix public role policy update payload

hanneskuettner committed:
Fix app access for users without role (which is a thing now apparently)

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

DanielBiegler reviewed (commented):
_( Just commenting for stuff I find along the way so we can check it off )_

Currently I cant attach users inside the roles view:

```
api dev: [15:22:10] PATCH /roles/43c9f03c-527a-4287-939b-baa46c6a0e70 500 46ms
api dev: [15:22:10.348] ERROR: select `admin_access` from `directus_roles` where `id` = '43c9f03c-527a-4287-939b-baa46c6a0e70' limit 1 - SQLITE_ERROR: no such column: admin_access
api dev:     err: {
api dev:       "type": "Error",
api dev:       "message": "select `admin_access` from `directus_roles` where `id` = '43c9f03c-527a-4287-939b-baa46c6a0e70' limit 1 - SQLITE_ERROR: no such column: admin_access",
api dev:       "stack":
api dev:           Error: select `admin_access` from `directus_roles` where `id` = '43c9f03c-527a-4287-939b-baa46c6a0e70' limit 1 - SQLITE_ERROR: no such column: admin_access
api dev:       "errno": 1,
api dev:       "code": "SQLITE_ERROR",
api dev:       "extensions": {
api dev:         "stack":
api dev:             Error: select `admin_access` from `directus_roles` where `id` = '43c9f03c-527a-4287-939b-baa46c6a0e70' limit 1 - SQLITE_ERROR: no such column: admin_access
api dev:       }
api dev:     }
```

This is because of `checkForOtherAdminUsers` inside of the `updateOne` of roles

```ts
	// TODO move to policies
	private async checkForOtherAdminUsers(
		key: PrimaryKey,
		users: Alterations<User, 'id'> | (string | Partial<User>)[],
	): Promise<void> {
		const role = await this.knex.select('admin_access').from('directus_roles').where('id', '=', key).first();
```

hanneskuettner commented:
Good catch, actually working on it as we speak :)

hanneskuettner committed:
Make sure that the /me endpoints always return minimal information, similar to /users/me

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

rijkvanzanten commented on app/src/interfaces/_system/system-permissions/system-permissions.vue:
```diff
@@ -0,0 +1,714 @@
+		}
+	}
+
+	function useUtil() {
```
Wondering if we should extract some of these useX functions to separate ts files to organize this massive setup function a bit

rijkvanzanten commented on app/src/modules/settings/components/navigation.vue:
```diff
@@ -27,9 +27,14 @@ const links = computed<Link[][]>(() => [
 			to: `/settings/roles`,
 		},
+		{
+			icon: 'badge',
```
Lets use `admin_panel_settings` as the icon here

```suggestion
			icon: 'admin_panel_settings',
```

rijkvanzanten commented on app/src/modules/settings/components/navigation.vue:
```diff
@@ -27,9 +27,14 @@ const links = computed<Link[][]>(() => [
 		},
 		{
 			icon: 'admin_panel_settings',
```
Lets use `group` here

rijkvanzanten commented on app/src/modules/settings/routes/policies/add-new.vue:
```diff
@@ -0,0 +1,97 @@
+
+const { saving, save } = useSave();
+
+function useSave() {
```
Lets extract this to a `use-save.ts`

rijkvanzanten commented on app/src/modules/settings/routes/policies/collection.vue:
```diff
@@ -0,0 +1,277 @@
+import { useRouter } from 'vue-router';
+import SettingsNavigation from '../../components/navigation.vue';
+
+// Taken from 20240328A-permissions-policies.ts
+const PUBLIC_POLICY_ID = 'abf8a154-5b1c-4a46-ac9c-7300570f4f17';
```
We should move / pull this from a single shared constant in `@directus/constants`

rijkvanzanten committed:
Tweak nav icons

rijkvanzanten committed:
Pull policy id from constants

rijkvanzanten committed:
Update permissions interface design to match

New design language in figma

hanneskuettner committed:
Some minor adjustments

- Make chip hover border more consistent
- Add "Remove" button to remove a full row of permissions, as in the UI mockup
- Fix table layout

hanneskuettner committed:
Clean up a few more things

hanneskuettner committed:
Fix `setFullAccess`

hanneskuettner committed:
Align collection view icons with navigation

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Don't query 'admin_access' for role

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Fix relation extraction and permissions for `$FOLLOW` fields

hanneskuettner committed:
Don't show `0 Items` for child rows, but `--` instead

hanneskuettner committed:
Make policy detail work in nested policy creating use case

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

hanneskuettner committed:
Merge branch 'auditus-21765' into auditus-app

rijkvanzanten committed:
Merge branch 'auditus-21765' into auditus-app

rijkvanzanten committed:
Remove unused v-icon override

rijkvanzanten reviewed (commented):
Code lgtm

hanneskuettner commented:
Any input on the open todos? (And still some todos left in code)

rijkvanzanten committed:
Move system collections to separate visual table

rijkvanzanten committed:
Navigate before refresh

Prevents a flash of the previous value to be visible in the table

rijkvanzanten commented:
@hanneskuettner Lets tackle shares in a different PR, as there's both app and API work to consider. Lets leave the types as is for now, we can always move them to `types` if we need to reuse them out of the app context

rijkvanzanten committed:
Move composable to separate file

qugkyp-vudgaz-2maQba reviewed (commented):
Looks great [](https://github.com/directus/directus/pull/22654/files)
